댓글 CRUD, 좋아요 기능 구현 및 예외처리

HaenaCho01 opened the pull request:
Commit 1. comment 기본 CRUD 기능 및 좋아요 기능
-기본적인 내용만 담아두었으므로 세부적인 내용은 다른 분들 코드 보면서 수정 및 조정해야 함
-좋아요 기능 취소가 가능해야 하므로 entity 새로 구성해야 할 듯 함 -> 이후 수정 및 개선 예정

Commit 2. 선택한 게시글에 대한 모든 댓글 조회 API 추가 및 postid를 Pathvariable로 받는 부분에 대한 코드 수정
-선택한 게시글 id를 Pathvariable로 받으면 해당 게시글에 대한 댓글을 모두 조회할 수 있도록 추가
(백엔드만 구현시 필요하지 않지 않을 것 같지만 프론트엔드를 구현시에 필요할 듯 하여 추가함)
-postid를 pathvariable 방식으로 받도록 코드 수정
-Pathvariable로 받은 posIid와 comment의 DB에 존재하는 postId가 일치하지 않다면 예외처리를 하도록 코드 수정
-예외처리 중 수정기능, 좋아요기능에서 Response에 message가 반환이 되지 않아 이후 수정 예정

Commit 3. 예외 처리 관련 코드 수정
-수정 기능, 좋아요 기능 예외처리 시 ResponseMessage가 반환이 안되는 문제 해결
-포스트 아이디나 댓글 아이디가 없는 번호일 때 예외처리 -> 아직 미해결. 추후 구현 예정

Commit 4. api 명세서에 맞춰 수정, User 기능 관련 작성된 코드에 맞춰 수정
-약속된 api 명세서와 다르게 잘못 작성되어 해당 내용 수정
-User와 연관된 코드를 User 브랜치에 작성된 코드에 맞춰 수정

Commit 5. ERD에 맞춘 코드 수정 및 API 명세서 수정에 따른 코드 재수정, PR에서 의논된 내용에 맞춰 코드 수정
-약속된 ERD와 다른 내용으로 작성된 코드가 있어 수정
-API 명세의 수정이 생겨 해당 내용에 맞게 코드 재수정
-PR에서 의논된 내용에 맞춰 코드 수정

HaenaCho01 committed:
해나 깃허브 pr테스트

HaenaCho01 committed:
해나 깃허브 pr테스트 삭제

HaenaCho01 committed:
comment 기능 init commit

HaenaCho01 committed:
comment 기본 CRUD 기능 및 좋아요 기능

-기본적인 내용만 담아두었으므로 세부적인 내용은 다른 분들 코드 보면서 수정 및 조정해야함
-좋아요 기능 취소가 가능해야하므로 entity 새로 구성해야 할 듯 함

HaenaCho01 committed:
Merge remote-tracking branch 'origin/comment-haena' into comment-haena

HaenaCho01 committed:
선택한 게시글에 대한 모든 댓글 조회 API 추가 및 postid를 Pathvariable로 받는 부분에 대한 코드 수정

-선택한 게시글 id를 Pathvariable로 받으면 해당 게시글에 대한 댓글을 모두 조회할 수 있도록 추가
(백엔드만 구현시 필요하지 않지 않을 것 같지만 프론트엔드를 구현시에 필요할 듯 하여 추가함)

-postid를 pathvariable 방식으로 받도록 코드 수정
-Pathvariable로 받은 postid와 comment의 DB에 존재하는 postid가 일치하지 않다면 예외처리를 하도록 코드 수정

-예외처리에 관한 핸들링이 잘못된 방법인 듯해 해당 부분 이후 수정 예정

HaenaCho01 committed:
예외 처리 관련 코드 수정

-좋아요 기능, 수정 기능 예외처리 시 ResponseMessage가 반환이 안되는 문제 해결

-포스트 아이디나 댓글 아이디가 없는 번호일 때 예외처리 -> 아직 미해결. 추후 구현 예정

HaenaCho01 commented:
예외 처리가 이 방법이 맞는지 잘 모르겠는데.. 아무리 공부해도 이해가 잘 안 가서 예외 처리 할 수 있는 분들 코드 올려주시면 해당 방법으로 수정 진행하도록 하겠습니다!

HaenaCho01 committed:
api 명세서에 맞춰 수정, User 기능 관련 작성된 코드에 맞춰 수정

-약속된 api 명세서와 다르게 잘못 작성되어 해당 내용 수정
-User와 연관된 코드를 User 브랜치에 작성된 코드에 맞춰 수정

HaenaCho01 committed:
주석 추가

-주석 추가

JisooPyo commented on src/main/java/com/sparta/trillionnewspeedproject/exception/GlobalExceptionHandler.java:
```diff
@@ -0,0 +1,15 @@
+import org.springframework.stereotype.Component;
+import org.springframework.web.bind.annotation.ExceptionHandler;
+
+@Component // @RestControllerAdvice 사용법을 아직 잘 모르겠음..
```
@RestControllerAdvice는 @ControllerAdvice + @ResponseBody

@ControllerAdvice는 @ExceptionHandler, @ModelAttribute, @InitBinder가 적용된 메서드들을 Controller 단에 적용하기 위해 고안된 어노테이션.
@ExceptionHandler : @Controller, @RestController에서만 적용가능한 예외처리 어노테이션
@InitBinder : Spring Validator를 사용시 @Valid 어노테이션으로 검증이 필요한 객체를 가져오기 전에 수행할 method를 지정해주는 어노테이션.

저도 구글링해서 확실히는 모르지만 검증했을 때의 예외를 처리하는데, 처리한 것을 객체로 돌려주고 싶을 때 쓰는 것 같아보여요! 

단순이 예외만 처리하고 싶다면 @ControllerAdvice를 적용하면 되고, 응답으로 객체를 리턴해야 한다면 @RestControllerAdvice를 적용하면 된다고 합니다!

적용범위를 클래스 단위나 패키지 단위로 제한할 수 있고, 아래와 같이 사용하면 된대요.

@RestControllerAdvice(basePackageClasses = (클래스명).class)
@RestControllerAdvice(basePackages = "com.sparta.....")

제가 너무 주르륵 받아 적기만 했는데
https://velog.io/@banjjoknim/RestControllerAdvice
여기 참고해보시면 좋을 거 같아요!

JisooPyo replied:
근데 아마 포스트맨으로 확인해봤을 때 잘 되었으니까..! 흠.. 이 부분은 정확히 튜터님에게 여쭤보고 쓰면 더 좋을 거 같아요!

HaenaCho01 replied:
오오 감사합니다! 붙여주신 블로그 참고하여 다시 공부해보겠습니다! @RestControllerAdivice를 쓰려고했더니 Exception을 파라미터로 사용하라고 해서 조금 어렵더라구요.. 아무래도 스프링 단이 아니라 자바 단의 예외 처리 쪽이 제가 스스로 느끼기에 많이 부족해서 거기부터 다시 공부해야할듯 합니다..! 시간될 때 튜터님 한번 찾아가봐야겠네요!! 감사합니다!!👍👍👍

JisooPyo replied:
가실 때 저랑 같이 가요!!

JisooPyo commented on src/main/java/com/sparta/trillionnewspeedproject/dto/CommentRequestDto.java:
```diff
@@ -0,0 +1,10 @@
+@Getter
+public class CommentRequestDto {
+	@NotBlank
+	private String comment;
```
이부분 Comment Entity와 약간 헷갈릴 수도 있을 것 같아서 commentContents 로 수정해주실 수 있을까요? 아마 refactor -> rename 하시면 comment 썼던 부분 다 한번에 바뀔거에요!

HaenaCho01 replied:
넵 수정 완료했습니다!

HaenaCho01 replied:
앗 contents로 수정했었는데 commentContents로 다시 수정해서 올려둘게요!

JisooPyo commented on src/main/java/com/sparta/trillionnewspeedproject/controller/CommentController.java:
```diff
@@ -0,0 +1,92 @@
+import java.util.List;
+
+@RestController
+@RequiredArgsConstructor
```
오 생성자를 어노테이션으로 대신한거군요! 머싯다!!

HaenaCho01 replied:
계속 만들자니 시간이 조금 걸리는 것 같아서 한번 사용해보았습니다ㅎㅎ

JisooPyo commented on src/main/java/com/sparta/trillionnewspeedproject/dto/CommentResponseDto.java:
```diff
@@ -0,0 +1,23 @@
+
+@Getter
+public class CommentResponseDto {
+	private Long id;
+	private String username;
+	private LocalDateTime createdAt;
+	private String comment;
+	private int likes;
```
추가 구현 기능에 댓글도 수정이 가능하게끔 되어 있어서 private LocalDateTime modifiedAt; 변수도 넣으면 좋을 거 같습니다!

HaenaCho01 replied:
넵 수정 완료했습니다!

JisooPyo commented on src/main/java/com/sparta/trillionnewspeedproject/entity/Comment.java:
```diff
@@ -0,0 +1,47 @@
+
+	@Id
+	@GeneratedValue(strategy = GenerationType.IDENTITY)
+	private Long id;
```
요거 commentId 로 변수명 설정하면 더 좋을 거 같아요..!

HaenaCho01 replied:
넵 수정 완료했습니다!

JisooPyo commented on src/main/java/com/sparta/trillionnewspeedproject/entity/Comment.java:
```diff
@@ -0,0 +1,47 @@
+	private String comment;
+
+	@ManyToOne(fetch = FetchType.LAZY)
+	@JoinColumn(name = "user_username", nullable = false)
```
name = "username"으로 하는것은 어떨까요?

HaenaCho01 replied:
ERD를 다시 보니 userId가 User의 PK고 저희가 해당 키로 연결하기로 했어서 해당 부분 name = "userId"로 수정 진행하겠습니다!

JisooPyo commented on src/main/java/com/sparta/trillionnewspeedproject/service/CommentService.java:
```diff
@@ -0,0 +1,124 @@
+
+		// postId 받은 것과 comment DB에 저장된 postId가 다를 경우 오류 코드 반환
+		if (postId != findComment(commentId).getPost().getId()) {
+			response.setStatus(404);
```
이렇게 하면 에러코드와 메시지가 담겨있는 json의 형태로 객체가 return되는 걸까요?

HaenaCho01 replied:
Service단에서 오류 상황별로 상태 코드를 컨트롤러 단에 넘겨주면 컨트롤러에서 바뀐 상태 코드를 확인하고 json 형태의 dto를 Response Body에 반환하는 방식으로 구현하였습니다!

JisooPyo replied:
그렇군요! 제가 아직까지 구조 파악이 어려워서 여쭤보았습니다! 설명 감사합니다!

JisooPyo commented on src/main/java/com/sparta/trillionnewspeedproject/service/CommentService.java:
```diff
@@ -0,0 +1,124 @@
+	}
+
+	// 선택한 댓글 좋아요 기능 추가
+	public CommentResponseDto commentLike(Long postId, Long commentId, User user, HttpServletResponse response) {
+
+		// postId 받은 것과 comment DB에 저장된 postId가 다를 경우 오류 코드 반환
+		if (postId != findComment(commentId).getPost().getId()) {
+			response.setStatus(404);
+			return null;
+
+		// 작성자가 좋아요를 시도할 경우 오류 코드 반환
+		} else if (checkUser(commentId, user)) {
+			response.setStatus(400);
+			return null;
+
+		// 오류가 나지 않을 경우 해당 댓글 좋아요 추가
+		} else {
+			Comment comment = findComment(commentId);
+			comment.updateLikes();
+			Comment savedComment = commentRepository.save(comment);
+			CommentResponseDto commentResponseDto = new CommentResponseDto(savedComment);
+			return commentResponseDto;
+		}
+	}
```
user가 여러 번 좋아요를 누르지 못하게 하려면 뭔가 user entity쪽에 boolean 값 likes 변수를 하나 주면 좋을까요..?

HaenaCho01 replied:
likes 를 누르면 DB에 User 정보가 저장되지 않고 지나가고 없어져서 boolean 값을 넣어도 비교할 대상이 없다 보니 해당 내용으로 구현하기에는 쪼오끔 어려움이 있지 않을까 싶긴 합니다 ㅠㅠ 저번에 이슈에서 의논한 내용으로 한번 구현 시도해보겠습니다!👍

JisooPyo replied:
https://velog.io/@korea3611/Spring-Boot%EA%B2%8C%EC%8B%9C%EA%B8%80-%EC%A2%8B%EC%95%84%EC%9A%94-%EA%B8%B0%EB%8A%A5-%EB%A7%8C%EB%93%A4%EA%B8%B0
이 블로그에서 보니까 Post의 like의 Entity를 아예 따로 만들고, postId, userid, likeId를 필드로 구성한 것 같습니다.
댓글의 좋아요 기능을 구현하려면 아예 CommentLike Entity를 아예 따로 만들고, likeId, postId, commentId, userId를 필드로 하여 만들어야 하는 것 같습니다!

HaenaCho01 committed:
ERD에 맞춘 코드 수정 및 API 명세서 수정에 따른 코드 수정

-약속된 ERD와 다른 내용으로 작성된 코드가 있어 수정
-API 명세의 수정이 생겨 해당 내용에 맞게 코드 수정

HaenaCho01 committed:
PR에서 의논된 내용에 맞춰 코드 수정

-Comment Entity의 멤버 comment -> commentContents
-modifiedAt 추가
-Comment Entity의 멤버 id -> commentId
-Comment Entity와 user 연결 이름 user_userId -> userId

JisooPyo reviewed (commented):
댓글 CRUD, 예외처리, 프론트단에 상태코드와 메시지 return해주는 것까지 완료!
좋아요 기능도 화이팅!!
